### ISSUE TITLE
Find and replace panel sometimes shows count of and highlighting of matches from a previous selection

### ISSUE THREAD

**Issue by WalterGR** [atom/find-and-replace#642]
Atom 1.3.2 on OS X 10.10.5.

I've just discovered this one, so I can't guarantee that this will reproduce 100%.

1.  Open a document in Atom.
2.  Perform a (non-regex) find and replace within a selection.
3.  In the editor, make another selection that contains one or more matches of the same "Find" text form step 2.
4.  Open Find and Replace via the keyboard binding for `find-and-replace:show-replace`.

Expected:

The count of matches of the Find text in the current selection is correct.

All matches in the selection are highlighted correctly in the editor.

Actual:

The count of matches may reflect the count of matches in a previous selection.

The matches that get highlighted in the editor may be from a previous selection.

Screenshots:

16 lines are selected, containing 16 matches of `;;`.  The Find and Replace panel reports that the count of matches is 13.  The editor has boxes around only 13 of the matches:

![atom-replace-with- in-selection -on-is-counting-and-showing-matches-from-a-previous-selection](https://cloud.githubusercontent.com/assets/380880/12252130/558db156-b889-11e5-96d2-c19988bb4daa.png)

13 lines are selected, containing 13 matches of `;;`.  The Find and Replace panel reports the count of matches in the entire document.  The editor has boxes around all matches in the document:

![atom-replace-with- in-selection -on-is-matching-in-the-entire-document](https://cloud.githubusercontent.com/assets/380880/12252158/8e5b5772-b889-11e5-9eb3-9e8d0386c9de.png)

**Comment by WalterGR** [comment]
Still repros in 1.4.0.

So, the title of this issue - "Find and Replace panel, editor sometimes show count of matches and highlighting of matches (respectively) from a previous selection" - isn't perfect, because in the following screenshot I had not selected the entire document previously, and yet it's showing the count of and boxes around matches in the entire document:

<img width="1097" alt="atom-replace-in-selection-matching-entire-document" src="https://cloud.githubusercontent.com/assets/380880/12406459/0b0690b6-be02-11e5-8d99-0659c7761003.png">

So a more accurate title would be more general.  Something like: "When Find pane is opened when text is selected, and 'Only In Selection' is already on, then the count of matches and boxes around matches may be incorrect".  I'll change it to that.

**Comment by 50Wliu** [comment]
@WalterGR That sounds similar to #298 - could you take a look please?  Maybe I was too quick in closing that issue.

**Comment by WalterGR** [comment]
@50Wliu: Yes, I think #298 subsumes this issue.  And the issues described definitely still repro.

Potentially with the exception of scenario 1:

> Press "Only in Selection"... also un-highlights the selection

That specific issue isn't reproducing for me right now in 1.4.0.  (I do know that 1.3.3 fixed an issue I opened: #631 - 'Toggling "Use Regex" clears selection in the editor'.  Maybe the issue was also fixed for toggling "Only in Selection".)

But the rest of the weirdness in #298 definitely still repros in 1.4.0.

**Comment by 50Wliu** [comment]
Ok, thanks.  If I remember I'll open up new issues for the rest of #298 tomorrow.

**Comment by WalterGR** [comment]
I'm glad I was cautious in my wording about scenario 1 in #298 ("Searching within current selection is very buggy".)  :)  Replying to my previous comment:

>  And the issues described in it definitely still repro.
> Potentially with the exception of scenario 1:
> > Press "Only in Selection"... also un-highlights the selection
> That specific issue isn't reproducing for me right now in 1.4.0. 

In fact, toggling "Only in Selection" _does_ un-highlight the selection if "Use Regex" is on.

And it turns out that #631 ('Toggling "Use Regex" clears selection in the editor') which I said was no longer reproing does repro - if "Only in Selection" is on.

Those points aren't relevant to this issue, but I wanted to correct what I've written earlier.  :)

I'll comment on #631 requesting to have it re-opened.

**Comment by WalterGR** [comment]
Additionally, when doing Replace All, it will replace the matches in whatever it thinks is the current region (selection or entire document) regardless of what the actual selection is.

**Comment by JosephTLyons** [comment]
My issue is that Atom still searches outside of the highlighted region.  Any word on this near 2 year old issue?

**Comment by Josh-West-Bose** [comment]
This is still an issue with 1.23.1

**Comment by JosephTLyons** [comment]
Crazy that this is still a bug.  Anyone have any word on it?
